### ISSUE TITLE
OSS-Fuzz frequent build failures

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#1444]
I get at least one email a week from OSS-Fuzz failing a build. These are intermittent failures as the next build usually succeeds and closes the issue. Their log doesn't give a whole lot of information about what's failing, other than a timeout:

```
Step #14: [1;93m[!] [1;97mWARNING: [0mTest case results in a timeout (skipping)[0m
Step #14: )B[?25h[0m[1;91m
Step #14: [-] PROGRAM ABORT : [0mAll test cases time out, giving up![1;91m
Step #14:          Location : [0mperform_dry_run(), src/afl-fuzz-init.c:1116
Step #14: 
Step #14: 
Step #14: BAD BUILD: fuzzing /tmp/not-out/zeek-pop3-fuzzer with afl-fuzz failed.
```

**Comment by jsiwek** [comment]
Guess it's related to this change which enabled AFL (see my comment): https://github.com/google/oss-fuzz/pull/5148

**Comment by jsiwek** [comment]
Potentially fixed now: https://github.com/google/oss-fuzz/pull/5148#issuecomment-800461864

**Comment by timwoj** [comment]
Looks like it's still a problem, unfortunately.

**Comment by Neverlord** [comment]
The issue boils down to a background thread accessing state that CAF cleaned up after leaving `main`. Usually this cannot happen, because CAF waits for all of its background threads in the destructor of `actor_system`.

However, it seems like the fuzzer setup does not clean up its state before leaving main. Hence, the destructor of the actor system does not run and the background threads happily carry on after returning from `main`. The regular Zeek `main` shuts down the Broker manager in `zeek::detail::cleanup`, so it seems like this only affects the fuzzer setup.

The recommend way in libfuzzer seems to be to wrap required state in a static object. But we have no way of forcing the runtime to destroy that static object *after* CAF destroys its (also static) [cleanup helper](https://github.com/actor-framework/actor-framework/blob/46a3c6a42e7ab5874930812f112bfe7aebf4c12e/libcaf_core/src/detail/meta_object.cpp#L42). So basically, that would be a trip to undefined behavior land. Is there a way to give the fuzzing library a piece of code that it runs before leaving `main`? If that isn't possible, then we could fix the issue in CAF by having each thread that CAF spawns holding on to a reference to the cleanup helper in a thread-local variable.

**Comment by Neverlord** [comment]
A patch is available here: https://github.com/actor-framework/actor-framework/pull/1242.

I can't confirm whether it's working, though. After installing a debug version of Zeek locally and setting `ZEEKPATH`, both fuzzer binaries crash for me:

```sh
$ ./src/fuzzers/zeek-pop3-fuzzer 
Standalone fuzzer processing 0 inputs
../../src/zeek/IntrusivePtr.h:157:10: runtime error: reference binding to null pointer of type 'zeek::Type'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../../src/zeek/IntrusivePtr.h:157:10 in 
Abort trap: 6

$ ./src/fuzzers/zeek-packet-fuzzer 
Standalone fuzzer processing 0 inputs
../../src/zeek/IntrusivePtr.h:157:10: runtime error: reference binding to null pointer of type 'zeek::Type'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../../src/zeek/IntrusivePtr.h:157:10 in 
Abort trap: 6
```

This is the stack trace I get from LLDB:

```
$ lldb ./src/fuzzers/zeek-packet-fuzzer 
(lldb) target create "./src/fuzzers/zeek-packet-fuzzer"
Current executable set to '/Users/neverlord/corelight/zeek/build/debug/src/fuzzers/zeek-packet-fuzzer' (x86_64).
(lldb) run
Process 42329 launched: '/Users/neverlord/corelight/zeek/build/debug/src/fuzzers/zeek-packet-fuzzer' (x86_64)
Standalone fuzzer processing 0 inputs
Process 42329 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = Null pointer use
    frame #0: 0x0000000115a5e4f0 libclang_rt.asan_osx_dynamic.dylib`__ubsan_on_report
libclang_rt.asan_osx_dynamic.dylib`__ubsan_on_report:
->  0x115a5e4f0 <+0>: pushq  %rbp
    0x115a5e4f1 <+1>: movq   %rsp, %rbp
    0x115a5e4f4 <+4>: popq   %rbp
    0x115a5e4f5 <+5>: retq   
Target 0: (zeek-packet-fuzzer) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = Null pointer use
  * frame #0: 0x0000000115a5e4f0 libclang_rt.asan_osx_dynamic.dylib`__ubsan_on_report
    frame #1: 0x0000000115a58fc8 libclang_rt.asan_osx_dynamic.dylib`__ubsan::Diag::~Diag() + 216
    frame #2: 0x0000000115a5a585 libclang_rt.asan_osx_dynamic.dylib`handleTypeMismatchImpl(__ubsan::TypeMismatchData*, unsigned long, __ubsan::ReportOptions) + 901
    frame #3: 0x0000000115a5a81c libclang_rt.asan_osx_dynamic.dylib`__ubsan_handle_type_mismatch_v1_abort + 60
    frame #4: 0x00000001005e7c98 libzeek_fuzzer_shared.dylib`zeek::IntrusivePtr<zeek::Type>::operator*(this=<unavailable>) const at IntrusivePtr.h:157:3 [opt]
    frame #5: 0x00000001005e5ead libzeek_fuzzer_shared.dylib`zeek::same_type(t1=<unavailable>, t2=0x00006070001ce760, is_init=<unavailable>, match_record_field_names=<unavailable>) at Type.h:763:24 [opt]
    frame #6: 0x0000000100c76604 libzeek_fuzzer_shared.dylib`zeek::same_type(arg_t1=<unavailable>, arg_t2=<unavailable>, is_init=<unavailable>, match_record_field_names=<unavailable>) at Type.cc:0 [opt]
    frame #7: 0x00000001005e5eda libzeek_fuzzer_shared.dylib`zeek::same_type(t1=<unavailable>, t2=0x00006070001ce6d0, is_init=false, match_record_field_names=true) at Type.h:763:14 [opt]
    frame #8: 0x00000001007d7163 libzeek_fuzzer_shared.dylib`zeek::detail::AssignExpr::TypeCheck(this=0x0000608000062120, attrs=<unavailable>) at Expr.cc:2413:9 [opt]
    frame #9: 0x00000001007d44cb libzeek_fuzzer_shared.dylib`zeek::detail::AssignExpr::AssignExpr(this=0x0000608000062120, arg_op1=<unavailable>, arg_op2=<unavailable>, arg_is_init=false, arg_val=zeek::ValPtr @ 0x00007ffeefbf67e0, attrs=<unavailable>, typecheck=<unavailable>) at Expr.cc:2326:10 [opt]
    frame #10: 0x00000001007d8a8c libzeek_fuzzer_shared.dylib`zeek::detail::AssignExpr::AssignExpr(this=<unavailable>, arg_op1=zeek::detail::ExprPtr @ 0x00007ffeefbf67a0, arg_op2=zeek::detail::ExprPtr @ 0x00007ffeefbf67c0, arg_is_init=<unavailable>, arg_val=zeek::ValPtr @ 0x00007ffeefbf67e0, attrs=0x00007ffeefbf6800, typecheck=<unavailable>) at Expr.cc:2304:2 [opt]
    frame #11: 0x00000001008237f0 libzeek_fuzzer_shared.dylib`zeek::IntrusivePtr<zeek::detail::AssignExpr> zeek::make_intrusive<zeek::detail::AssignExpr, zeek::IntrusivePtr<zeek::detail::Expr>, zeek::IntrusivePtr<zeek::detail::Expr>, bool&>(args=<unavailable>, args=<unavailable>, args=<unavailable>) at IntrusivePtr.h:186:26 [opt]
    frame #12: 0x0000000100821069 libzeek_fuzzer_shared.dylib`zeek::detail::get_assign_expr(op1=zeek::detail::ExprPtr @ 0x00007ffeefbf9460, op2=zeek::detail::ExprPtr @ 0x00007ffeefbf9480, is_init=false) at Expr.cc:5137:10 [opt]
    frame #13: 0x00000001001aa06f libzeek_fuzzer_shared.dylib`yyparse() at parse.y:510:19 [opt]
    frame #14: 0x000000010053ce7a libzeek_fuzzer_shared.dylib`zeek::detail::setup(argc=<unavailable>, argv=<unavailable>, zopts=<unavailable>) at zeek-setup.cc:629:2 [opt]
    frame #15: 0x0000000100004ff8 zeek-packet-fuzzer`::LLVMFuzzerInitialize(argc=<unavailable>, argv=<unavailable>) at fuzzer-setup.h:42:7 [opt]
    frame #16: 0x0000000100013b0b zeek-packet-fuzzer`main(argc=<unavailable>, argv=0x00007ffeefbff750) at standalone-driver.cc:21:2 [opt]
    frame #17: 0x00007fff2033a621 libdyld.dylib`start + 1
```

Is something wrong/missing in my setup?

**Comment by jsiwek** [comment]
Note there's two separate things going on with OSS-Fuzz:
* This issue, which is a failure to even get past OSS-Fuzz build-phase
* #1349: the spurious/noisy heap-use-after-free issues reported by OSS-Fuzz in Zeek shutdown logic

The following is more related to the later and should probably followup further from that separate issue:

> Is there a way to give the fuzzing library a piece of code that it runs before leaving `main`?

Not that I found, but also mentioned in #1349, that's the kind of thing I had wanted to use to address it.

> A patch is available here: [actor-framework/actor-framework#1242](https://github.com/actor-framework/actor-framework/pull/1242).

Thanks, I think that may work for this particular heap-use-after-free problem, though there's still the underlying inability to call `zeek::detail::cleanup()` inside the fuzzing scaffold that may have easy potential for other problems/regressions of this nature (not an argument against incremental/partial/workaround improvements, just noting it).

> I can't confirm whether it's working, though. After installing a debug version of Zeek locally and setting `ZEEKPATH`, both fuzzer binaries crash for me:

OSS-Fuzz doesn't currently use UBSan, so not using it in your testing may be the simplest answer.

The longer answer for running w/ UBSan is that Zeek hasn't resolved all UB in the default suite of sanitizer-checks (e.g. #1175, #1176), so there's a way to disable those specific ones in the meantime (which is what Zeek CI does): https://github.com/zeek/zeek/wiki/Checking-for-Memory-Errors-and-Leaks#configure-zeek-with-undefinedbehaviorsanitizer

**Comment by Neverlord** [comment]
Thanks for this summary! So at least it's not my setup. 🙂

When building without UBSan, I see `Could not create Broker master store` errors, but at least ASAN isn't complaining.

Just as an aside: even if we could pass something to the fuzzer that it runs before leaving main, I think the CAF patch is still the best solution. Just waiting on all actors isn't enough and trying to wait on all CAF threads plus the actor system destructor from the outside gets tricky.

**Comment by jsiwek** [comment]
The AFL fuzz engine got disabled for now via https://github.com/google/oss-fuzz/pull/5671 to get OSS-Fuzz builds going again.
